Optimize Node::getParent and usage in previews

juliusknorr opened the pull request:
`Node::getParent` does not need to query the database again if the parent node was already used before for constructing the child node. In those cases it is now passed to the child node and can be reused.

The second commit avoids using the getParent call on preview generation which saves one query.

Found in Collectives where there are similar optimisations, but this would be a generic use to help prevent this for users of getParent()

juliusknorr committed:
Avoid extra query for the parent node if we can pass it along

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Save getting the parent node when generating the preview

Signed-off-by: Julius Härtl <jus@bitgrid.net>

CarlSchwan reviewed (approved):
Clever :)

CarlSchwan commented:
One failure left

> 1) TrashbinTest::testExpireOldFilesUtilLimitsAreMet
> Failed asserting that 0 is identical to 10.
>
> /home/runner/work/server/server/apps/files_trashbin/tests/TrashbinTest.php:366

juliusknorr committed:
Only pass parent if paths match

As the user folder might be initialized by the root from two levels
down the hierarchy, passing this as a parent only works if the path matches

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliushaertl commented:
Tests revealed an additional issue for cases where we cannot just pass the current node as parent, e.g. if `get('user_id/files')` is called on the root folder, as then the root would be considered the parent (while `/user_id` would be the actual one). Pushed a possible fix, so let's see what tests tell now.

CarlSchwan commented:
The only faikure now is the one about session that is unrelated

juliushaertl commented:
Fix for the failing session test is in https://github.com/nextcloud/server/pull/33668
